### PR TITLE
Federate /v2/observation entity expression expansions

### PR DESCRIPTION
**Pull request by clincoln8**
## Goal
Implemented a mechanism to support dynamic entity discovery from a remote mixer (or any configured source) when using the Spanner backend for `/v2/observation` calls with entity expressions. This ensures that when a user queries for observations using an expression like `geoId/06<-containedInPlace+{typeOf:County}`, we can fetch data for entities that might only exist in the remote mixer's graph.

## Requirements based on Legacy Flow
Based on the legacy Cloud SQL flow, we needed to achieve the following:
*   **Original Expression to Remote**: The Remote Mixer must receive the original request with the expression intact so it can perform its own expansion against its complete graph.
*   **Remote-Only Expansion**: We should call the Remote Mixer to resolve the expression to a list of child entities.
*   **Federated Resolution for Local**: Local sources (like Spanner) must be queried with the resolved list of DCIDs (including entities found in remote), and the expression must be cleared to prevent them from attempting expansion they cannot support.
*   **Optimal Fallback**: When no remote mixer is configured, the Spanner data source must use the current optimal one-shot query (`GetObservationsContainedInPlace`).

## Implementation Details
Chose a design that splits the responsibility between a new Dispatcher Processor and the `SpannerDataSource`:

### 1. `RelationExpressionProcessor` (in `internal/server/dispatcher`)
*   **Why here?**: This processor acts as a coordinating middleware in the Dispatcher layer, executed BEFORE the dispatcher fans out requests to individual data sources. Putting it here makes it backend-agnostic and available for other backends in the future.
*   **What it does**: Checks if the request is an observation request with an expression. If so, calls the configured source to expand the expression and stores the resulting DCIDs in the context using the key `RelationExpressionExpandedEntities`.
*   **Resilience**: If the call to the source fails, it logs a warning and continues, allowing fallback to local-only execution.

### 2. `SpannerDataSource` (in `internal/server/spanner`)
*   **What it does**: In the `Observation` method, it delegates fetching to `fetchObservations` helper. If `fetchObservations` finds pre-expanded entities in the context, it calls `expandAndMergeEntities` to fetch local child places from Spanner (using `sds.Node` which handles optimizations like `linkedContainedInPlace`), merges them with the pre-expanded DCIDs, and queries Spanner for observations of all entities in this merged list. If not present, it falls back to the optimized one-shot query (`GetObservationsContainedInPlace`).

This design keeps the request object immutable in the processor and avoids passing a large list of DCIDs over the network to the Remote Mixer for the observation call.

### Unit Tests
Added comprehensive unit tests to verify this flow:
*   **`relation_expression_processor_test.go`**: Tests that the processor correctly calls the source and populates the context, handles no-ops for different request types, and handles failures by falling back gracefully.
*   **`datasource_test.go`**: Tests that `SpannerDataSource` correctly uses the context data when present, falls back to one-shot query when missing, and handles standard queries without expressions correctly.

### Manual Verification
To verify this change manually:
*   Started up a local Mixer with a Spanner backend and a Remote Mixer configured.
*   Called the `/v2/observation` endpoint with a `containedInPlace` expression query (`geoId/06<-containedInPlace+{typeOf:County}`).
*   Verified that the response correctly merged results, containing both the local dummy county (`geoId/06999`) and all the standard remote counties from `api.datacommons.org`.
*   Verified that the local data was preferred and merged correctly according to the design.

### CODE REVIEW

File: internal/server/dispatcher/relation_expression_processor.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+) ([]string, error) {
+	property := fmt.Sprintf("<-containedInPlace+{typeOf:%s}", containedInPlace.ChildPlaceType)
+	nodeReq := &pbv2.NodeRequest{
+		Nodes:    []string{containedInPlace.Ancestor},
+		Property: property,
+	}
+
+	// Call Node API on the remote source directly.
+	// Pass 0 for pageSize to match legacy flow (ignored by RemoteDataSource.Node).
+	resp, err := remoteSource.Node(ctx, nodeReq, 0)
```
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/dispatcher/relation_expression_processor.go]: Will handle in a follow-up PR to iterate all pages from node response in RemoteDatasource

File: cmd/main.go [hunk 2]
```diff
@@ -423,19 +425,26 @@ func main() {
 			var redisProcessor dispatcher.Processor = redis.NewCacheProcessor(redisClient)
 			processors = append(processors, &redisProcessor)
 		}
-		slog.Info("After Redis setup")
+
+		// Relation Expression Processor
+		if remoteDataSource != nil {
+			slog.Info("remoteDataSource is configured, setting up relation expression processor")
+			var relationExpressionProcessor dispatcher.Processor = dispatcher.NewRelationExpressionProcessor(dataSources)
```
Comment (keyurva) [inline cmd/main.go]: Chatted offline. Very cool that we're injecting this only when there is a remote data source. We'll further make the change to only pass the remote data source to the processor instead of all data sources.

File: internal/server/dispatcher/relation_expression_processor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/dispatcher/relation_expression_processor.go]: Decided to handle this in a follow up PR so ok to go with this for now.

File: internal/server/dispatcher/relation_expression_processor.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	containedInPlace, err := v2.ParseContainedInPlace(expr)
+	if err != nil {
+		slog.Error("RelationExpressionProcessor: failed to parse expression", "expression", expr, "error", err)
+		return Continue, fmt.Errorf("failed to parse expression: %w", err)
+	}
+
+	// Fetch entities from remote source.
+	dcids, err := p.fetchRemoteEntities(rc.Context, remoteSource, containedInPlace)
+	if err != nil {
+		return Continue, err
```
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/dispatcher/relation_expression_processor.go]: @keyurva Let me know what you think the behavior here should be.

File: internal/server/dispatcher/relation_expression_processor_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		mockNodeResponse   *pbv2.NodeResponse
+		mockNodeError      error
+		expectedOutcome    Outcome
+		expectedErr        bool
+		expectNodeCalled   bool
+		expectedContextKey bool
+		expectedDCIDs      []string
+	}{
+		{
+			name:        "Happy Path: Expression + Remote Mixer",
```
Comment (keyurva) [inline internal/server/dispatcher/relation_expression_processor_test.go:55]: Let's trim the number of test cases to what helps functional coverage and is easier to maintain.

File: internal/server/spanner/golden/datasource_test.go [hunk 5]
```diff
@@ -585,3 +588,163 @@ func TestBulkVariableGroupInfo_Filtering(t *testing.T) {
 		t.Errorf("Expected ExcludedNode in response, but it was missing")
 	}
 }
+
+func TestSpannerObservation_ExpressionExpansion(t *testing.T) {
```
Comment (keyurva) [inline internal/server/spanner/golden/datasource_test.go:593]: Not for this PR but since this file is meant to be for goldens, let's add a TODO to move all unit tests to a separate test file.

File: internal/server/spanner/datasource.go [hunk 6]
```diff
@@ -196,9 +198,37 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
 		if err != nil {
 			return nil, fmt.Errorf("error getting observations (contained in): %v", err)
 		}
-		observations, err = sds.client.GetObservationsContainedInPlace(ctx, variables, containedInPlace)
-		if err != nil {
-			return nil, fmt.Errorf("error getting observations (contained in): %v", err)
+
+		// Check if expanded DCIDs are available in context (passed by preprocessor).
+		if remoteDcids, ok := ctx.Value(dispatcher.RelationExpressionExpandedEntities).([]string); ok {
+			slog.Debug("SpannerDataSource: using expanded remote entities from context", "count", len(remoteDcids))
```
Comment (keyurva) [inline internal/server/spanner/datasource.go]: nit: ok for this to be info level since we're only logging the count.

File: internal/server/spanner/datasource.go [hunk 7]
```diff
@@ -212,6 +242,32 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
...
+// fetchLocalChildPlaces fetches child places from Spanner.
+// Example:
+//   Inputs: ancestor="geoId/06", childType="County"
+//   Outputs: []string{"geoId/06001", "geoId/06002"}, nil
+func (sds *SpannerDataSource) fetchLocalChildPlaces(ctx context.Context, ancestor, childType string) ([]string, error) {
+	arc := &v2.Arc{
+		Out:        false, // <-
+		SingleProp: v2.ContainedInPlaceProperty,
+		Filter:     map[string][]string{"typeOf": {childType}},
+	}
```
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/spanner/datasource.go]: I will update this to call sds.Node instead of using the lower level calls.

File: internal/server/spanner/datasource.go [hunk 8]
```diff
@@ -196,9 +198,37 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
...
+			slog.Debug("SpannerDataSource: fetched local child places", "count", len(localDcids))
+
+			// 2. Merge local and remote DCIDs.
+			mergedDcids := append(localDcids, remoteDcids...)
+			slices.Sort(mergedDcids)
+			mergedDcids = slices.Compact(mergedDcids)
+
+			// 3. Fetch observations for the merged list.
+			if len(mergedDcids) == 0 {
+				return &pbv2.ObservationResponse{}, nil
```
Comment (keyurva) [inline internal/server/spanner/datasource.go]: nit: Can add a log here indicating no child dcids were found.

File: internal/server/spanner/datasource.go [hunk 9]
```diff
@@ -196,9 +198,37 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
 		if err != nil {
 			return nil, fmt.Errorf("error getting observations (contained in): %v", err)
 		}
-		observations, err = sds.client.GetObservationsContainedInPlace(ctx, variables, containedInPlace)
-		if err != nil {
-			return nil, fmt.Errorf("error getting observations (contained in): %v", err)
+
+		// Check if expanded DCIDs are available in context (passed by preprocessor).
+		if remoteDcids, ok := ctx.Value(dispatcher.RelationExpressionExpandedEntities).([]string); ok {
```
Comment (keyurva) [inline internal/server/spanner/datasource.go]: nit: Consider handling expanded dcids in a helper method for readability.

File: internal/server/spanner/datasource.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/spanner/datasource.go]: Is recursive lookup required here?

File: internal/server/spanner/datasource.go [hunk 10]
```diff
@@ -212,6 +242,32 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
...
+// Example:
+//   Inputs: ancestor="geoId/06", childType="County"
+//   Outputs: []string{"geoId/06001", "geoId/06002"}, nil
+func (sds *SpannerDataSource) fetchLocalChildPlaces(ctx context.Context, ancestor, childType string) ([]string, error) {
+	arc := &v2.Arc{
+		Out:        false, // <-
+		SingleProp: v2.ContainedInPlaceProperty,
+		Filter:     map[string][]string{"typeOf": {childType}},
+	}
+	localEdges, err := sds.client.GetNodeEdgesByID(ctx, []string{ancestor}, arc, 1000, 0)
```
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/spanner/datasource.go]: Can handle pagination separately along with the pagination in the processor.

**Review by keyurva — APPROVED** [review]
Thanks!

File: internal/server/dispatcher/relation_expression_processor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/dispatcher/relation_expression_processor.go]: The gemini suggestion (logging + fallback) makes sense.

File: cmd/main.go (hunk 2 above)
Comment (clincoln8) [reply cmd/main.go]: brilliant, done!

File: internal/server/dispatcher/relation_expression_processor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/dispatcher/relation_expression_processor.go]: done

File: internal/server/dispatcher/relation_expression_processor_test.go (hunk 4 above)
Comment (clincoln8) [reply internal/server/dispatcher/relation_expression_processor_test.go:55]: reduced by one, kept the rest to test edge cases/ maintain high code-path-coverage. added clearer comments about what is being tested.

File: internal/server/spanner/golden/datasource_test.go (hunk 5 above)
Comment (clincoln8) [reply internal/server/spanner/golden/datasource_test.go:593]: added a todo

File: internal/server/spanner/datasource.go (hunk 9 above)
Comment (clincoln8) [reply internal/server/spanner/datasource.go]: done

File: internal/server/spanner/datasource.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/spanner/datasource.go]: yes. updated to call Node with `<-containedInPlace+`

File: internal/server/spanner/datasource.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/spanner/datasource.go]: added a todo

File: internal/server/dispatcher/relation_expression_processor.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	property := fmt.Sprintf("<-containedInPlace+{typeOf:%s}", containedInPlace.ChildPlaceType)
+	nodeReq := &pbv2.NodeRequest{
+		Nodes:    []string{containedInPlace.Ancestor},
+		Property: property,
+	}
+
+	// Call Node API on the source directly.
+	// PageSize (0) is ignored for now.
+	// TODO: iterate paged responses to compile full entity set
+	resp, err := source.Node(ctx, nodeReq, 0)
```
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/dispatcher/relation_expression_processor.go:102]: will handle pagination in follow up

File: internal/server/spanner/datasource.go [hunk 12]
```diff
@@ -191,27 +192,109 @@ func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.Observa
...
+func (sds *SpannerDataSource) fetchLocalChildPlaces(ctx context.Context, ancestor, childType string) ([]string, error) {
+	property := fmt.Sprintf("<-%s+{typeOf:%s}", v2.ContainedInPlaceProperty, childType)
+	nodeReq := &pbv2.NodeRequest{
+		Nodes:    []string{ancestor},
+		Property: property,
+	}
+
+	// Call Node API to let it handle optimizations (e.g., mapping to linkedContainedInPlace).
+	// TODO: Handle iterating paged response if NextToken is present.
+	resp, err := sds.Node(ctx, nodeReq, 1000)
```
[the remark this answers is not in the document]
Comment (clincoln8) [reply internal/server/spanner/datasource.go:279]: will handle pagination in follow up
